varinfo/utilities.py

```diff
@@ -121,20 +125,126 @@ def get_xml_attribute(
 
     if attribute_element is not None:
         value_type = attribute_element.get('type', 'String')
-        numpy_type = DAP4_TO_NUMPY_MAP.get(value_type, str)
-
-        value_elements = attribute_element.findall(f'{namespace}Value')
 
-        if len(value_elements) > 1:
-            attribute_value = [
-                numpy_type(value_element.text) for value_element in value_elements
-            ]
-        elif len(value_elements) == 1:
-            attribute_value = numpy_type(value_elements[0].text)
+        if value_type != 'Container':
+            attribute_value = get_xml_attribute_value(
+                attribute_element,
+                namespace,
+                value_type,
+                default_value,
+            )
         else:
-            attribute_value = default_value
+            attribute_value = get_xml_container_attribute(attribute_element, namespace)
+
+    else:
+        attribute_value = default_value
+
+    return attribute_value
+
+
+def get_xml_attribute_value(
+    attribute_element: Element,
+    namespace: str,
+    value_type: str,
+    default_value: Any | None = None,
+) -> Any | None:
+    """Extract the value (single or list) for an XML attribute. If there are
+    no attributes matching the required name, then return the supplied default
+    value. If no default value is supplied, the default used is `None`.
 
+    """
+    numpy_type = DAP4_TO_NUMPY_MAP.get(value_type, str)
+
+    value_elements = attribute_element.findall(f'{namespace}Value')
+
+    if len(value_elements) > 1:
+        attribute_value = [
+            numpy_type(value_element.text) for value_element in value_elements
+        ]
+    elif len(value_elements) == 1:
+        attribute_value = numpy_type(value_elements[0].text)
     else:
         attribute_value = default_value
 
     return attribute_value
+
+
+def get_xml_container_attribute(
+    container_element: Element, namespace: str
+) -> dict[str, Any | None]:
+    """Extract a dictionary of attribute values when an attribute is a container
+    for further attributes. This function is recursive, and so nested containers
+    will be treated in the same way.
+
+    """
+    attribute_dictionary = {}
+
+    for child in list(container_element):
+        child_name = child.get('name')
+        child_type = child.get('type', 'String')
+
+        if child_type != 'Container':
+            attribute_dictionary[child_name] = get_xml_attribute_value(
+                child,
+                namespace,
+                child_type,
+            )
+        else:
+            attribute_dictionary[child_name] = get_xml_container_attribute(
+                child,
+                namespace,
+            )
+
+    return attribute_dictionary
+
+
+def get_nested_xml_attribute(
```

Comment:
Maybe not well described... retrieving a nested XML attribute is happening when trying to get an attribute buried away on a group or variable in a DMR. The attribute container stuff above is separate, in that everything has a single parent attribute element, but there are nested attributes/values within that single attribute. (Hopefully that didn't just make things sound more confusing)